File: app/src/main/java/com/project/scratchstudio/kith_andoid/UI/ProfileFragment.java
package com.project.scratchstudio.kith_andoid.UI;

import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.project.scratchstudio.kith_andoid.Activities.EditActivity;
import com.project.scratchstudio.kith_andoid.Activities.HomeActivity;
import com.project.scratchstudio.kith_andoid.Activities.ProfileActivity;
import com.project.scratchstudio.kith_andoid.R;
import com.project.scratchstudio.kith_andoid.Service.PicassoCircleTransformation;
import com.project.scratchstudio.kith_andoid.UserPresenter;
import com.project.scratchstudio.kith_andoid.app.BaseFragment;
import com.project.scratchstudio.kith_andoid.app.Const;
import com.project.scratchstudio.kith_andoid.databinding.FragmentProfileBinding;
import com.project.scratchstudio.kith_andoid.network.ApiClient;
import com.project.scratchstudio.kith_andoid.network.apiService.UserApi;
import com.project.scratchstudio.kith_andoid.network.model.referral.ReferralResponse;
import com.project.scratchstudio.kith_andoid.network.model.user.User;
import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.Picasso;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public class ProfileFragment extends BaseFragment {

    private boolean anoutherUser;
    private User user;
    private UserApi userApi;
    private CompositeDisposable disposable = new CompositeDisposable();
    private View share;
    private FragmentProfileBinding binding;

    private UserPresenter userPresenter;

    public static ProfileFragment newInstance(Bundle bundle) {
        ProfileFragment fragment = new ProfileFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Bundle bundle = getArguments();
        user = (User) bundle.getSerializable("user");
        anoutherUser = bundle.getBoolean("another_user");
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_profile, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        userPresenter = new UserPresenter(getContext());
        userApi = ApiClient.getClient(getContext()).create(UserApi.class);

        if (anoutherUser) {
            binding.edit.setVisibility(View.GONE);
            binding.exit.setVisibility(View.GONE);

            binding.myShare.setVisibility(View.GONE);
            binding.share.setVisibility(View.VISIBLE);
        }

        fullField();
        initButtonListener();
    }

    private void fullField() {

        if (user.getUrl() != null) {
            Picasso.with(getContext()).load(user.getUrl().replaceAll("@[0-9]*", ""))
                    .placeholder(R.mipmap.person)
                    .error(R.mipmap.person)
                    .transform(new PicassoCircleTransformation())
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(binding.photo);
        }

        binding.name.setText(user.getFirstName());
        binding.surname.setText(user.getLastName());
        binding.phone.setText(user.getPhone());
        if (user.getMiddleName() != null && !user.getMiddleName().equals("") && !user.getMiddleName().toLowerCase().equals("null")) {
            binding.middlename.setText(user.getMiddleName());
        } else {
            binding.middlename.setText("-");
        }
        binding.position.setText(user.getPosition());
        binding.usersCount.setText(String.valueOf(user.getUsersCount()));
        if (user.getDescription() == null || user.getDescription().equals("null") || user.getDescription().equals("")) {
            binding.description.setText("-");
        } else {
            binding.description.setText(user.getDescription());
        }

        if (user.getEmail() == null || user.getEmail().equals("null") || user.getEmail().equals("")) {
            binding.email.setText("-");
        } else {
            binding.email.setText(user.getEmail());
        }
    }

    private boolean isNetworkConnected() {
        ConnectivityManager cm = (ConnectivityManager) getActivity().getSystemService(Context.CONNECTIVITY_SERVICE);
        return (cm != null && cm.getActiveNetworkInfo() != null);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == 0) {
            share.setEnabled(true);
        }
        if (requestCode == 1) {
            if (data != null && resultCode == 0) {
                if (isNetworkConnected()) {
                    refreshUser();
                } else {
                    Toast.makeText(getContext(), "Нет подключения к интернету", Toast.LENGTH_SHORT).show();
                }
            }
        }
    }

    public void refreshUser() {
        User user = HomeActivity.getMainUser();
        if (user.getUrl() != null) {
            Picasso.with(getContext()).load(user.getUrl().replaceAll("@[0-9]*", ""))
                    .placeholder(R.mipmap.person)
                    .error(R.mipmap.person)
                    .transform(new PicassoCircleTransformation())
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(binding.photo);
        }
        binding.name.setText(user.getFirstName());
        binding.surname.setText(user.getLastName());
        binding.phone.setText(user.getPhone());
        if (user.getMiddleName() != null && !user.getMiddleName().equals("") && !user.getMiddleName().toLowerCase().equals("null")) {
            binding.middlename.setText(user.getMiddleName());
        } else {
            binding.middlename.setText("-");
        }
        binding.position.setText(user.getPosition());
        binding.usersCount.setText(String.valueOf(user.getUsersCount()));
        if (user.getDescription() == null || user.getDescription().equals("") || user.getDescription().toLowerCase().equals("null")) {
            binding.description.setText("-");
        } else {
            binding.description.setText(user.getDescription());
        }
        if (user.getEmail() == null || user.getEmail().equals("null") || user.getEmail().equals("")) {
            binding.email.setText("-");
        } else {
            binding.email.setText(user.getEmail());
        }
    }

    private void initButtonListener(){
        binding.back.setOnClickListener(this::onClickBack);
        binding.exit.setOnClickListener(this::onClickExitButton);
        binding.edit.setOnClickListener(this::onClickEditButton);
        binding.share.setOnClickListener(this::onClickContactShare);
        binding.myShare.setOnClickListener(this::onClickContactShare);
        binding.phone.setOnClickListener(this::onClickCallPhone);
    }

    public void onClickBack(View view) {
        ((HomeActivity) getActivity()).backFragment();
    }

    private void onClickExitButton(View view) {
        view.setEnabled(false);
        ((HomeActivity) getActivity()).exit();
    }

    private void onClickEditButton(View view) {
        view.setEnabled(false);
        Intent intent = new Intent(getActivity(), EditActivity.class);
        startActivityForResult(intent, 1);
        view.setEnabled(true);
    }

    private void onClickCallPhone(View view) {
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setData(Uri.parse("tel:" + user.getPhone()));
        startActivity(intent);
    }

    private void onClickContactShare(View view) {
        view.setEnabled(false);
        share = view;
        disposable.add(
                userApi.getReferralCode(HomeActivity.getMainUser().id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<ReferralResponse>() {
                            @Override
                            public void onSuccess(ReferralResponse response) {
                                Log.i("REf", response.getStatus() + " " + response.getReferral());
                                String result = "ФИО: " + binding.surname.getText() + " " + binding.name.getText() + " " + binding.middlename.getText() + "\n"
                                        + (binding.phone.getText().length() > 1 ? "Телефон: " + binding.phone.getText() + "\n" : "")
                                        + (binding.email.getText().length() > 1 ? "Эл.почта: " + binding.email.getText() + "\n" : "")
                                        + (user.position.length() > 1 ? "Сфера деятельности: " + user.position + "\n" : "")
                                        + "Приложение: " + Const.APP_URL + "\n"
                                        + "Реф.код для регистрации: " + response.getReferral() + "\n"
                                        + getString(R.string.signature);

                                shareDate(result);
                            }

                            @Override
                            public void onError(Throwable e) {
                                String result = "ФИО: " + binding.surname.getText() + " " + binding.name.getText() + " " + binding.middlename.getText() + "\n"
                                        + "Телефон: " + binding.phone.getText() + "\n"
                                        + "Эл.почта: " + binding.email.getText() + "\n"
                                        + "Сфера деятельности: " + user.position + "\n"
                                        + "Приложение: " + Const.APP_URL + "\n"
                                        + getString(R.string.signature);

                                shareDate(result);
                            }
                        })
        );
    }

    private void shareDate(String share) {
        Intent sendIntent = new Intent();
        sendIntent.setAction(Intent.ACTION_SEND);
        sendIntent.putExtra(Intent.EXTRA_TEXT, share);
        sendIntent.setType("text/plain");
        startActivityForResult(sendIntent, 0);
    }

    @Override
    public boolean onBackPressed() {
        return ((HomeActivity) getActivity()).backFragment();
    }

}

File: app/src/main/java/com/project/scratchstudio/kith_andoid/ui/home_package/user_profile/ProfileFragment.kt
package com.project.scratchstudio.kith_andoid.ui.home_package.user_profile

import android.accounts.NetworkErrorException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.activityViewModels
import androidx.navigation.findNavController
import com.project.scratchstudio.kith_andoid.R
import com.project.scratchstudio.kith_andoid.UserPresenter
import com.project.scratchstudio.kith_andoid.activities.HomeActivity
import com.project.scratchstudio.kith_andoid.app.BaseFragment
import com.project.scratchstudio.kith_andoid.app.Const
import com.project.scratchstudio.kith_andoid.app.UserType
import com.project.scratchstudio.kith_andoid.databinding.FragmentProfileBinding
import com.project.scratchstudio.kith_andoid.model.UserModelView
import com.project.scratchstudio.kith_andoid.network.model.user.User
import com.project.scratchstudio.kith_andoid.service.PicassoCircleTransformation
import com.project.scratchstudio.kith_andoid.ui.home_package.user_edit_profile.EditProfileFragment
import com.project.scratchstudio.kith_andoid.view_model.CurrentUserViewModel
import com.project.scratchstudio.kith_andoid.view_model.MainUserViewModel
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.Picasso
import java.util.Locale

class ProfileFragment : BaseFragment() {

    private var share: View? = null
    private var binding: FragmentProfileBinding? = null
    private var userPresenter: UserPresenter? = null

    private val currentUserViewModel: CurrentUserViewModel by activityViewModels()
    private val mainUserViewModel: MainUserViewModel by activityViewModels()
    private lateinit var user: UserModelView
    private lateinit var userType: UserType

    override fun onCreate(savedInstanceState: Bundle?) {
        setHasOptionsMenu(true)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_profile, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_profile_exit -> {
                HomeActivity.exit(activity!!)
                return true
            }
            R.id.menu_profile_edit -> {
                activity!!.findNavController(R.id.nav_host_fragment_home)
                        .navigate(ProfileFragmentDirections.actionProfileFragmentToEditProfileFragment())
                return true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentProfileBinding.inflate(inflater, container, false)
        userType = ProfileFragmentArgs.fromBundle(arguments!!).userType
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        userPresenter = UserPresenter(context!!)
        user = if(userType == UserType.MAIN_USER) {
            mainUserViewModel.getMainUser().value!!
        } else {
            currentUserViewModel.getCurrentUser().value!!
        }
        fullField()
        initButtonListener()
    }

    private fun fullField() {

        if (user.photo != null && user.photo != "") {
            Picasso.with(context).load(user.photo!!.replace("@[0-9]*".toRegex(), ""))
                    .placeholder(R.mipmap.person)
                    .error(R.mipmap.person)
                    .transform(PicassoCircleTransformation())
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(binding!!.photo)
        }

        binding!!.name.text = user.firstName
        binding!!.surname.text = user.lastName
        binding!!.phone.text = user.phone
        if (user.middleName != "" && user.middleName != "" && user.middleName.toLowerCase(Locale.getDefault()) != "null") {
            binding!!.middlename.text = user.middleName
        } else {
            binding!!.middlename.text = "-"
        }
        binding!!.position.text = user.position
        binding!!.usersCount.text = user.usersCount.toString()
        if (user.description == "" || user.description == "null" || user.description == "") {
            binding!!.description.text = "-"
        } else {
            binding!!.description.text = user.description
        }

        if (user.email == "" || user.email == "null" || user.email == "") {
            binding!!.email.text = "-"
        } else {
            binding!!.email.text = user.email
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == 0) {
            share!!.isEnabled = true
        }
        if (requestCode == 1) {
            if (data != null && resultCode == 0) {
                userPresenter!!.getUser(object : UserPresenter.GetUserCallback {
                    override fun onSuccess(userResponse: User) {
                        if (userResponse.photo != "") {
                            user.photo = userResponse.photo
                            Log.i("Profile", userResponse.photo)
                        } else {
                            Log.i("Profile", "null photo")
                        }
                        refreshUser(userResponse)
                        Toast.makeText(context, "Данные обновлены", Toast.LENGTH_SHORT).show()
                    }

                    override fun onError(networkError: NetworkErrorException) {
                        Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                    }
                }, user.id)
            }
        }
    }

    private fun refreshUser(refUser: User) {
        if (refUser.photo != "" && refUser.photo != null) {
            Picasso.with(context).load(refUser.photo!!.replace("@[0-9]*".toRegex(), ""))
                    .placeholder(R.mipmap.person)
                    .error(R.mipmap.person)
                    .transform(PicassoCircleTransformation())
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(binding!!.photo)
        }
        binding!!.name.text = refUser.firstName
        binding!!.surname.text = refUser.lastName
        binding!!.phone.text = refUser.phone
        if (refUser.middleName != "" && refUser.middleName != "" && refUser.middleName.toLowerCase(Locale.getDefault()) != "null") {
            binding!!.middlename.text = refUser.middleName
        } else {
            binding!!.middlename.text = "-"
        }
        binding!!.position.text = refUser.position
        binding!!.usersCount.text = refUser.usersCount.toString()
        if (refUser.description == "" || refUser.description == "" || refUser.description.toLowerCase(Locale.getDefault()) == "null") {
            binding!!.description.text = "-"
        } else {
            binding!!.description.text = refUser.description
        }
        if (refUser.email == "" || refUser.email == "null" || refUser.email == "") {
            binding!!.email.text = "-"
        } else {
            binding!!.email.text = refUser.email
        }
    }

    private fun initButtonListener() {
//        binding!!.share.setOnClickListener { this.onClickContactShare(it) }
        binding!!.myShare.setOnClickListener { this.onClickContactShare(it) }
        binding!!.phone.setOnClickListener { this.onClickCallPhone(it) }
    }

    private fun onClickBack(view: View) {
        (activity as HomeActivity).backFragment()
    }

    private fun onClickEditButton(view: View) {
        view.isEnabled = false
        val intent = Intent(activity, EditProfileFragment::class.java)
        startActivityForResult(intent, 1)
        view.isEnabled = true
    }

    private fun onClickCallPhone(view: View) {
        val intent = Intent(Intent.ACTION_DIAL)
        intent.data = Uri.parse("tel:" + user.phone)
        startActivity(intent)
    }

    private fun onClickContactShare(view: View) {
        view.isEnabled = false
        share = view
        val result = ("ФИО: " + binding!!.surname.text + " " + binding!!.name.text + " " + binding!!.middlename
                .text + "\n"
                + (if (binding!!.phone.text.length > 1) "Телефон: " + binding!!.phone.text + "\n" else "")
                + (if (binding!!.email.text.length > 1) "Эл.почта: " + binding!!.email.text + "\n" else "")
                + (if (user.position.length > 1) "Сфера деятельности: " + user.position + "\n" else "")
                + "Приложение: " + Const.APP_URL + "\n"
                + "Реф.код для регистрации: " + user.referralCode + "\n"
                + getString(R.string.signature))
        shareDate(result)
    }

    private fun shareDate(share: String) {
        val sendIntent = Intent()
        sendIntent.action = Intent.ACTION_SEND
        sendIntent.putExtra(Intent.EXTRA_TEXT, share)
        sendIntent.type = "text/plain"
        startActivityForResult(sendIntent, 0)
    }

    override fun onBackPressed(): Boolean {
        return (activity as HomeActivity).backFragment()
    }

    companion object {

        fun newInstance(bundle: Bundle): ProfileFragment {
            val fragment = ProfileFragment()
            fragment.arguments = bundle
            return fragment
        }
    }

}
